updates in mediapipe docs (#2166)
openvinotoolkit/model_server: docs/mediapipe.md

@endsphinxdirective

## Introduction

> MediaPipe is an open-source framework for building pipelines to perform inference over arbitrary sensory data. If comes with a wide range of calculators/nodes which can be applied for unlimited number of scenarios in image and media analytics, generative AI, transformers and many more. Here can be found more information about [MediaPipe framework ](https://developers.google.com/mediapipe/framework/framework_concepts/overview)

Reviewer: ```suggestion
MediaPipe is an open-source framework for building pipelines to perform inference over arbitrary sensory data. It comes with a wide range of calculators/nodes which can be applied for unlimited number of scenarios in image and media analytics, generative AI, transformers and many more. Here can be found more information about [MediaPipe framework ](https://developers.google.com/mediapipe/framework/framework_concepts/overview)
```

![mp_graph_modes](./mp_graph_modes.png)

Check [our MediaPipe github fork](https://github.com/openvinotoolkit/mediapipe) to learn how to use the OpenVINO calculator inside standalone MediaPipe application. 

This guide gives information about:

* <a href="#ovms-calculators">OpenVINO Model Server Calculators</a>
* <a href="#create-graph">How to create the graph for deployment in OpenVINO Model Server</a>
* <a href="#graph-deploy">Graph deployment</a>
* <a href="#testing">Deployment testing</a>
* <a href="#client">Using MediaPipe graphs from the remote client </a>
* <a href="#updating-graph">How to update existing graphs to use OV for inference </a>
* <a href="#adding-calculator">Adding your own MediaPipe calculator to OpenVINO Model Server </a>
* <a href="#graph-examples">Demos and examples</a>
* <a href="#current-limitations">Current Limitations</a>

## OpenVINO calculators <a name="ovms-calculators"></a>

We are introducing a set of calculators which can bring to the graphs execution the advantage of OpenVINO Runtime.

Check their [documentation](https://github.com/openvinotoolkit/mediapipe/blob/main/mediapipe/calculators/ovms)

> ## Python calculator

Reviewer: It will be quite unhandy to use our PythonExecutorCalculator outside OVMS as of now.

Included OpenVINO inference session calculator should include the reference to the model name as configured in `subconfig.json`.
Here is example of the `subconfig.json`:
```json
{
    "model_config_list": [
        {"config": {
                "name": "model1_name",
                "base_path": "model1"
        	  }
	      }
    ]
}
```

### Starting OpenVINO Model Server with Mediapipe servables
MediaPipe servables configuration is to be placed in the same json file like the 
[models config file](starting_server.md).
While models are defined in section `model_config_list`, graphs are configured in
the `mediapipe_config_list` section. 

When the MediaPipe graphs artifacts are packaged like presented above, configuring the OpenVINO Model Server is very easy. Just a `config.json` needs to be prepared with a list of all the graphs to be deployed:
```json
{
    "model_config_list": [],
    "mediapipe_config_list": [
    {
        "name":"mediapipe_graph_name"
    },
    {
        "name":"mediapipe2",
        "base_path":"non_default_path"
    }
    ]
}
```

> Nodes in the MediaPipe graphs can reference both to the models configured in model_config_list section and in subconfigs.

Reviewer: ```suggestion
Nodes in the MediaPipe graphs can reference both the models configured in model_config_list section and in subconfigs.
```

mediapipe_graph_name/
├── graph.pbtxt
├── model1
│   └── 1
│       ├── model.bin
│       └── model.xml
├── model2
│   └── 1
│       ├── model.bin
│       └── model.xml
└── subconfig.json
```

The `graph.pbtxt` should include the definition of the mediapipe graph. 
The `subconfig.json` is an extension to the main model server `config.json` configuration file. The subconfig should include the definition of the models used in the graph nodes. They will be loaded during the model server initialization and ready to use when the graph starts executing.
Included OpenVINO inference session calculator should include the reference to the model name as configured in `subconfig.json`.
Here is example of the `subconfig.json`:
```json
{
    "model_config_list": [
        {"config": {
                "name": "model1_name",
                "base_path": "model1"
        	  }
	      }
    ]
}
```

### Starting OpenVINO Model Server with Mediapipe servables
MediaPipe servables configuration is to be placed in the same json file like the 
[models config file](starting_server.md).
While models are defined in section `model_config_list`, graphs are configured in
the `mediapipe_config_list` section. 

> When the mediapipe graphs artifacts are packaged like presented above, configuring the OpenVINO Model Server is very easy. Just a `config.json` needs to be prepared with a list of all the graphs to be deployed:

Reviewer 2: ```suggestion
When the MediaPipe graphs artifacts are packaged like presented above, configuring the OpenVINO Model Server is very easy. Just a `config.json` needs to be prepared with a list of all the graphs to be deployed:
```

```json
{
    "model_config_list": [],
    "mediapipe_config_list": [
    {
        "name":"mediapipe_graph_name"
    },
    {
        "name":"mediapipe2",
        "base_path":"non_default_path"
    }
    ]
}
```

> Nodes in the Mediapipe graphs can reference both to the models configured in model_config_list section and in subconfigs.

Reviewer 2: ```suggestion
Nodes in the MediaPipe graphs can reference both to the models configured in model_config_list section and in subconfigs.
```

There is also an option to avoid any data conversions in the serialization and deserialization by the OpenVINO Model Server. When the input stream is of type REQUEST, it will be passed-through to the calculator. The receiving calculator will be in charge of deserializing it and interpreting all the content. Likewise, the output format RESPONSE delegate to the calculator creating a complete KServe response message to the client. That gives extra flexibility in the data format.

**Note:** For list of supported packet types and tags of `OpenVINOInferenceCalculator`, including graph examples, check documentation of [OpenVINO Model Server calculators](https://github.com/openvinotoolkit/mediapipe/blob/main/mediapipe/calculators/ovms/).

### Side packets
Side packets are special parameters which can be passed to the calculators at the beginning of the graph initialization. It can tune the behavior of the calculator like set the object detection threshold or number of objects to process.

> With KServe gRPC API you are also able to push side input packets into graph. They are to be passed as KServe request parameters. They can be of type `string`, `int64` or `boolean`.

Reviewer: Shall we say how to access them "on the other side" or we don't want to get into details in this document?

mediapipe_graph_name/
├── graph.pbtxt
├── model1
│   └── 1
│       ├── model.bin
│       └── model.xml
├── model2
│   └── 1
│       ├── model.bin
│       └── model.xml
└── subconfig.json
```

The `graph.pbtxt` should include the definition of the mediapipe graph. 
The `subconfig.json` is an extension to the main model server `config.json` configuration file. The subconfig should include the definition of the models used in the graph nodes. They will be loaded during the model server initialization and ready to use when the graph starts executing.
Included OpenVINO inference session calculator should include the reference to the model name as configured in `subconfig.json`.
Here is example of the `subconfig.json`:
```json
{
    "model_config_list": [
        {"config": {
                "name": "model1_name",
                "base_path": "model1"
        	  }
	      }
    ]
}
```

### Starting OpenVINO Model Server with Mediapipe servables
MediaPipe servables configuration is to be placed in the same json file like the 
[models config file](starting_server.md).
While models are defined in section `model_config_list`, graphs are configured in
the `mediapipe_config_list` section. 

> When the mediapipe graphs artifacts are packaged like presented above, configuring the OpenVINO Model Server is very easy. Just a `config.json` needs to be prepared list all the graphs for deployments:

Reviewer: ```suggestion
When the mediapipe graphs artifacts are packaged like presented above, configuring the OpenVINO Model Server is very easy. Just a `config.json` needs to be prepared with a list of all the graphs to be deployed:
```

Subconfig file may only contain *model_config_list* section  - in the same format as in [models config file](starting_server.md).

## Deployment testing <a name="testing"></a>
### Debug logs
The simples method to validate the graph execution is to set the Model Server `log_level` do `DEBUG`.
`docker run --rm -it -v $(pwd):/config openvino/model_server:latest --config_path /config/config.json --log_level DEBUG`

> It will report in a verbose way all the operations in the mediapipe framework from the graph initialization and also the graphs execution.

Reviewer: ```suggestion
It will report in a verbose way all the operations in the mediapipe framework from the graph initialization and execution.
```

> After the model server, you could confirm with the graph has correct format and all the required models are loaded successfully.

Reviewer: That line is quite messy. Not sure what we want to say here.

> Note that graph loading is not confirming if all the calculators are compiled into the model server build. That can be confirmed with after sending the request to the KServe endpoint.

Reviewer: ```suggestion
Note that graph loading is not confirming if all the calculators are compiled into the model server build. That can be confirmed after sending the request to the KServe endpoint.
```

Is it true? Will the graph load if there's an issue with the calculator? I suppose if the graph loads and gets AVAILABLE, we know that calculators are available.

### Tracing
Currently the graph tracing on the model server side is not supported. If you would like to take advantage of mediapipe tracing to identify the graph bottleneck, test the graph from the mediapipe application level. Build an example application similarly to [holistic app](https://github.com/openvinotoolkit/mediapipe/tree/main/mediapipe/examples/desktop/holistic_tracking) with the steps documented on [mediapipe tracking](https://github.com/google/mediapipe/blob/master/docs/tools/tracing_and_profiling.md).

### Benchmarking
While you implemented and deployed the graph you have several options to test the performance.
To validate the throughput for unary requests you can use the [benchmark client](../demos/benchmark/python#mediapipe-benchmarking).

For streaming gRPC connections, there is available [rtps_client](https://github.com/openvinotoolkit/model_server/tree/llama-ovms/demos/mediapipe/holistic_tracking#rtsp-client).
It can generate the load to gRPC stream and the mediapipe graph based on the content from RTSP video stream, MPG4 file or from the local camera.

## Using MediaPipe graphs from the remote client <a name="client"></a>

> MediaPipe graphs can use the same gRPC KServe Inference API both the the unary calls and the streaming. 

Reviewer: ```suggestion
MediaPipe graphs can use the same gRPC KServe Inference API both for the unary calls and the streaming. 
```

> The same client libraries with KServe API support can be used in both cases. The client code for the unary and streaming is a bit different.

Reviewer: a bit? 😄

Check the [code snippets](https://docs.openvino.ai/2023.2/ovms_docs_clients_kfs.html)

Review also the information about the [gRPC streaming feature](./streaming_endpoints.md)

Graphs can be queried for their state using the calls [GetModelStatus](model_server_grpc_api_kfs.md), [REST Model Status](model_server_rest_api_kfs.md) 
and [GetModelMetadata](model_server_grpc_api_kfs.md) and [REST Model Metadata](model_server_rest_api_kfs.md).

The difference in using the MediaPipe graphs and individual models is in version management. In all calls to the MediaPipe graphs,
the version parameter is ignored. MediaPipe graphs are not versioned. Though, they can reference a particular version of the models in the graph.

## How to update existing graphs to use OV for inference <a name="updating-graph"></a>

> If you would like to reusing existing graph and replace Tensorflow execution with OpenVINO backend, check this guide TBD.

Reviewer: ```suggestion
If you would like to reuse existing graph and replace Tensorflow execution with OpenVINO backend, check this guide TBD.
```

Subconfig file may only contain *model_config_list* section  - in the same format as in [models config file](starting_server.md).

## Deployment testing <a name="testing"></a>
### Debug logs

> The simples method to validate the graph execution is to set the Model Server `log_level` to `DEBUG`.

Reviewer: ```suggestion
The simplest method to validate the graph execution is to set the Model Server `log_level` to `DEBUG`.
```

`docker run --rm -it -v $(pwd):/config openvino/model_server:latest --config_path /config/config.json --log_level TRACE`

It will report in a verbose way all the operations in the mediapipe framework from the graph initialization and execution.
The model server logs could confirm the graph correct format and loading all the required models.
Note that graph definition loading is not confirming if all the calculators are compiled into the model server. That can be tested after sending the request to the KServe endpoint.
During the requests processing, the logs will include info about calculators initialization and processing the nodes.

### Tracing

> Currently the graph tracing on the model server side is not supported. If you would like to take advantage of mediapipe tracing to identify the graph bottleneck, test the graph from the mediapipe application level. Build an example application similarly to [holistic app](https://github.com/openvinotoolkit/mediapipe/tree/main/mediapipe/examples/desktop/holistic_tracking) with the steps documented on [mediapipe tracking](https://github.com/google/mediapipe/blob/master/docs/tools/tracing_and_profiling.md).

Reviewer: Is it possible to reference openvino fork in both links?

Subconfig file may only contain *model_config_list* section  - in the same format as in [models config file](starting_server.md).

## Deployment testing <a name="testing"></a>
### Debug logs

> The simples method to validate the graph execution is to set the Model Server `log_level` do `TRACE`.

Reviewer 2: ```suggestion
The simples method to validate the graph execution is to set the Model Server `log_level` to `DEBUG`.
```

Reviewer 2: https://github.com/openvinotoolkit/model_server/blob/main/src/logging.cpp#L120

Reviewer 2: But we should change it to TRACE at some point I think.

`docker run --rm -it -v $(pwd):/config openvino/model_server:latest --config_path /config/config.json --log_level TRACE`

It will report in a verbose way all the operations in the mediapipe framework from the graph initialization and execution.
The model server logs could confirm the graph correct format and loading all the required models.
Note that graph definition loading is not confirming if all the calculators are compiled into the model server. That can be tested after sending the request to the KServe endpoint.

> During the requests processing, the logs will include info about calculators registration and processing the nodes.

Reviewer 2: ```suggestion
During the requests processing, the logs will include infoabout calculators registration and processing the nodes.
```
There is no info that calculator is registered - only in case of error we expect message from MP to inform us that sth is missing.

> During the requests processing, the logs will include info about calculators registration and processing the nodes.

Reviewer 2: ```suggestion
During the requests processing, the logs will include info about calculators initialization and processing the nodes.
```

## Using MediaPipe graphs from the remote client <a name="client"></a>

MediaPipe graphs can use the same gRPC KServe Inference API both for the unary calls and the streaming. 
The same client libraries with KServe API support can be used in both cases. The client code for the unary and streaming is different. 
Check the [code snippets](https://docs.openvino.ai/2023.2/ovms_docs_clients_kfs.html)

Review also the information about the [gRPC streaming feature](./streaming_endpoints.md)

Graphs can be queried for their state using the calls [GetModelStatus](model_server_grpc_api_kfs.md), [REST Model Status](model_server_rest_api_kfs.md) 
and [GetModelMetadata](model_server_grpc_api_kfs.md) and [REST Model Metadata](model_server_rest_api_kfs.md).

The difference in using the MediaPipe graphs and individual models is in version management. In all calls to the MediaPipe graphs,
the version parameter is ignored. MediaPipe graphs are not versioned. Though, they can reference a particular version of the models in the graph.

## How to update existing graphs to use OV for inference <a name="updating-graph"></a>

> If you would like to reuse existing graph and replace Tensorflow execution with OpenVINO backend, check this guide TBD.

Reviewer 2: ```suggestion
If you would like to reuse existing graph and replace Tensorflow execution with OpenVINO backend, check [this guide](mediapipe_conversion.md)
```

```

 Make sure the REGISTER_CALCULATOR(your_calculator); macro is present in the calculator file that you have added.

## MediaPipe Graphs Examples <a name="graphs-examples"></a>

[Holistic analysis](../demos/mediapipe/holistic_tracking)

[Image classification](../demos/mediapipe/image_classification/README.md)

[Object detection](../demos/mediapipe/object_detection)

[Multi model](../demos/mediapipe/multi_model_graph/README.md)

## Current limitations <a name="current-limitations"></a>
- MediaPipe graphs are supported only for gRPC KServe API.

- KServe ModelMetadata call response contains only input and output names. In the response shapes will be empty and datatypes will be `"INVALID"`.

> - Binary inputs are not supported for MediaPipe graphs.

Reviewer: Well, depends what binary inputs mean here. For passthrough request it's possible. And with Python calc it will also work. So maybe we should mention that binary inputs for OV Tensor type of inputs are not supported?
